In [2]:
#importamos librerias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
datadf=pd.read_excel("Gastos y costos 20-23.xlsx",skiprows=5,sheet_name='2020')
datadf.head(3)

FECHA                   FOLIO                                  UUID  \
0 2020-01-01                   39408  BD20F436-72CC-4BE2-AD37-5999D71AA391   
1 2020-01-01  0001010100750847116902  85AEAC2E-6395-44A0-9CB9-754D06893375   
2 2020-01-01  0001010100752163714102  BD0DFE85-D776-41CB-9C5C-4FA7923E687A   

            RFC                       PROVEEDOR         TIPO GASTO GASTO  \
0  ARE080115PBA             ARSI RENTA SA DE CV         MAQUINARIA   NaN   
1  BNM840515VB1  BANCO NACIONAL DE MEXICO, S.A.  COMISION BANCARIA   NaN   
2  BNM840515VB1  BANCO NACIONAL DE MEXICO, S.A.  COMISION BANCARIA   NaN   

                                         DESCRIPCION   TC  IMPORTE      IVA  \
0  Descripción: REVOLVEDORA, Modelo: ULTRAMP9A, M...  1.0      0.0  2248.28   
1                                   COMISION COBRADA  1.0    490.0    78.40   
2                                   COMISION COBRADA  1.0    195.0    31.20   

   RET ISR  RET IVA  TOTAL MX  TOTAL SAT TIPO   STATUS POLIZA  Columna1  
0      0.0      0.0   2248.28    16300.0    I  Vigente    NaN  14051.72  
1      0.0      0.0    568.40      568.4    I  Vigente    NaN      0.00  
2      0.0      0.0    226.20      226.2    I  Vigente    NaN      0.00

In [4]:
#Comenzamos la limpieza de la primer hoja, realizando una copia de nuestro data frame primero
df1=datadf.copy()

Podemos observar que se presenta un gran número de valores faltantes, así que debemos optar por algún método de sustitución que nos ayude a rellenar estos sesgos sin afectar demasido los datos


In [6]:
vn=df1.isnull().sum()
vn

FECHA             0
FOLIO           189
UUID              0
RFC               0
PROVEEDOR         0
TIPO GASTO        0
GASTO          2502
DESCRIPCION       0
TC              391
IMPORTE          34
IVA             268
RET ISR           0
RET IVA           0
TOTAL MX          0
TOTAL SAT         0
TIPO              1
STATUS            0
POLIZA         3321
Columna1          0
dtype: int64

***Para limpiar la columna folio, con 189 valores nulos, se opta por usar el método de String, dado que el folio sirve como identificador único, no podemos arriesgarnos a asignar datos repetidos, además, de esta forma, se podrán filtrar más fácil los valores faltantes***

In [7]:
df1["FOLIO"]=df1["FOLIO"].fillna("N/A")

In [8]:
#Corroboramos que se llenen exitosamente
vn=df1.isnull().sum()
vn


FECHA             0
FOLIO             0
UUID              0
RFC               0
PROVEEDOR         0
TIPO GASTO        0
GASTO          2502
DESCRIPCION       0
TC              391
IMPORTE          34
IVA             268
RET ISR           0
RET IVA           0
TOTAL MX          0
TOTAL SAT         0
TIPO              1
STATUS            0
POLIZA         3321
Columna1          0
dtype: int64

***Ahora, para la columna gasto, se hará una sustitución con String, dado que la variable es de tipo texto*** 


In [9]:
df1["GASTO"]=df1["GASTO"].fillna("SIN DEFINIR")

In [10]:
#Corroboramos
vn=df1.isnull().sum()
vn

FECHA             0
FOLIO             0
UUID              0
RFC               0
PROVEEDOR         0
TIPO GASTO        0
GASTO             0
DESCRIPCION       0
TC              391
IMPORTE          34
IVA             268
RET ISR           0
RET IVA           0
TOTAL MX          0
TOTAL SAT         0
TIPO              1
STATUS            0
POLIZA         3321
Columna1          0
dtype: int64

***Ahora, para la columna TC, dado que se tiene el mismo valor en cada espacio no nulo, se puede usar el tercer método de sustitución visto en clase, el de rellenar con un número en concreto, en este caso, el 1, dado que es el valor que está siempre presente, se puede asumir que es el valor faltante***

In [12]:
df1["TC"]=df1["TC"].fillna(1)

In [13]:
#Corroboramos
vn=df1.isnull().sum()
vn

FECHA             0
FOLIO             0
UUID              0
RFC               0
PROVEEDOR         0
TIPO GASTO        0
GASTO             0
DESCRIPCION       0
TC                0
IMPORTE          34
IVA             268
RET ISR           0
RET IVA           0
TOTAL MX          0
TOTAL SAT         0
TIPO              1
STATUS            0
POLIZA         3321
Columna1          0
dtype: int64

***En cuanto a la columna IMPORTE, se puede llevar a cabo una sustitución por el método mean, para poder conservar la media del conjunto de datos original*** 

In [14]:
df1["IMPORTE"]=df1["IMPORTE"].fillna(round(datadf["IMPORTE"].mean(),1))


In [15]:
#Verificamos
vn=df1.isnull().sum()
vn

FECHA             0
FOLIO             0
UUID              0
RFC               0
PROVEEDOR         0
TIPO GASTO        0
GASTO             0
DESCRIPCION       0
TC                0
IMPORTE           0
IVA             268
RET ISR           0
RET IVA           0
TOTAL MX          0
TOTAL SAT         0
TIPO              1
STATUS            0
POLIZA         3321
Columna1          0
dtype: int64

***En la columna IVA, se decide hacer lo mismo que en la columna IMPORTE, para poder conservar la media del conjunto de datos original***

In [17]:
df1["IVA"]=df1["IVA"].fillna(round(datadf["IVA"].mean(),1))
vn=df1.isnull().sum()
vn


FECHA             0
FOLIO             0
UUID              0
RFC               0
PROVEEDOR         0
TIPO GASTO        0
GASTO             0
DESCRIPCION       0
TC                0
IMPORTE           0
IVA               0
RET ISR           0
RET IVA           0
TOTAL MX          0
TOTAL SAT         0
TIPO              1
STATUS            0
POLIZA         3321
Columna1          0
dtype: int64

***Para la columna TIPO, se hace una sustitución usando el método bfill, para rellenarlo con el valor anterior al faltante***

In [19]:
df1["TIPO"]=df1["TIPO"].fillna(method="bfill")
vn=df1.isnull().sum()
vn

/var/folders/19/7c2smbhs6b3g5ffy1ph3r__m0000gn/T/ipykernel_51221/2472197598.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df1["TIPO"]=df1["TIPO"].fillna(method="bfill")


FECHA             0
FOLIO             0
UUID              0
RFC               0
PROVEEDOR         0
TIPO GASTO        0
GASTO             0
DESCRIPCION       0
TC                0
IMPORTE           0
IVA               0
RET ISR           0
RET IVA           0
TOTAL MX          0
TOTAL SAT         0
TIPO              0
STATUS            0
POLIZA         3321
Columna1          0
dtype: int64

***Finalmente, para las pólizas faltantes, se asume que el cliente no cuenta con póliza, por lo que se hará uso del método de sustitución de string***

In [21]:
df1["POLIZA"]=df1["POLIZA"].fillna("N/A")
vn=df1.isnull().sum()
vn

FECHA          0
FOLIO          0
UUID           0
RFC            0
PROVEEDOR      0
TIPO GASTO     0
GASTO          0
DESCRIPCION    0
TC             0
IMPORTE        0
IVA            0
RET ISR        0
RET IVA        0
TOTAL MX       0
TOTAL SAT      0
TIPO           0
STATUS         0
POLIZA         0
Columna1       0
dtype: int64

In [22]:
#Guardamos nuestro archivo csv
df1.to_csv("GyC21.csv")